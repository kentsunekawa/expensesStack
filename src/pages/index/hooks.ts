import { useCallback, useMemo } from 'react'
import { makeVar, useReactiveVar } from '@apollo/client'

import {
  dateToString,
  utcDateStringToDate,
  getfirstDateOfAfterNthMonth,
  getlastDateOfAfterNthMonth,
} from 'src/utils'
import { Stage } from 'src/operations/types.d'
import { useGetExpensesLazyQuery } from 'src/operations/queries/__generated__/GetExpenses'

type SearchQuery = {
  date: Date
  category: { id: string; name: string } | null
}

const searchQueryVar = makeVar<SearchQuery>({
  date: new Date(),
  category: null,
})

export const changeSearchQuery = (searchQuery: Partial<SearchQuery>) =>
  searchQueryVar({ ...searchQueryVar(), ...searchQuery })

export const useSearchQuery = () => ({
  searchQuery: useReactiveVar(searchQueryVar),
})

export const useGetExpenses = () => {
  const [getExpenses, { data, loading, error }] = useGetExpensesLazyQuery()

  const doGetExpenses = useCallback(
    (searchQuery: SearchQuery) => {
      void getExpenses({
        variables: {
          stage: import.meta.env.PROD ? Stage.Published : Stage.Draft,
          where: {
            date_gte: dateToString(
              getfirstDateOfAfterNthMonth(searchQuery.date, 0),
              'yyyy-MM-dd',
            ),
            date_lte: dateToString(
              getlastDateOfAfterNthMonth(searchQuery.date, 0),
              'yyyy-MM-dd',
            ),
            category: searchQuery.category
              ? {
                  id: searchQuery.category?.id,
                }
              : undefined,
          },
        },
      })
    },
    [getExpenses],
  )

  const expenses = useMemo(
    () =>
      data
        ? data.expenses.map(({ id, amount, memo, date, category }) => ({
            id,
            amount,
            memo,
            date: utcDateStringToDate(date, 'Date'),
            category: category
              ? {
                  id: category.id,
                  name: category.name,
                  color: (category.color?.hex ?? '') as string,
                }
              : null,
          }))
        : null,
    [data],
  )

  const fetchStatus = useMemo(
    () => ({
      isLoading: loading,
      isError: !!error,
    }),
    [loading, error],
  )

  return {
    doGetExpenses,
    expenses,
    fetchStatus,
  }
}
