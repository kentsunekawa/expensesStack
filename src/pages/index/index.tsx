import { useState, useEffect } from 'react'

import { toggleIsShowExpenseRegister } from 'src/hooks'

const Home: React.FC = () => {
  const [isOpen, setIsOpen] = useState<boolean>(true)

  useEffect(() => {
    toggleIsShowExpenseRegister(true)
  }, [])

  return <>Home</>
}

export default Home
