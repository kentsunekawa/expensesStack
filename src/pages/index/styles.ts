// import from libraries
import { css } from 'styled-components'

// import from this project
import { StyleBaseData } from 'src/hooks'

export const createStyles = ({ theme }: StyleBaseData) => ({
  container: css``,
  goToRegiserButtonPosi: css`
    position: fixed;
    right: 16px;
    bottom: 72px;
  `,
  goToRegiserButton: css`
    width: 56px;
    height: 56px;
    background: ${theme.palette.primary.main};
    color: '#fff';
  `,
  totalDisplayArea: css`
    position: fixed;
    left: 0;
    bottom: 0;
    width: 100%;
  `,
})
