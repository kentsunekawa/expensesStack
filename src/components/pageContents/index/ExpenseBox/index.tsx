// import from libraries
import 'styled-components/macro'
import { Paper } from '@mui/material'
import { getDate } from 'date-fns'

// import from this project
import { dateToString } from 'src/utils'
import { Expense } from 'src/types'
import { useStyle } from 'src/hooks'
import { Heading, Text } from 'src/components/parts/Texts'
import { createStyles } from './styles'

export type Props = {
  expense: Expense
}

export const ExpenseBox: React.FC<Props> = ({
  expense: { amount, category, date },
}) => {
  const { styles } = useStyle(createStyles, {
    color: category.color,
  })

  return (
    <Paper css={styles.container} variant='elevation'>
      <div css={styles.left}>
        <div css={styles.dateArea.container}>
          <p css={styles.dateArea.dateText}>{getDate(date)}</p>
          <p css={styles.dateArea.dayText}>{dateToString(date, 'E')}</p>
        </div>
        <div css={styles.iconArea.container}>
          <Text textAlign='left'>{category.name}</Text>
        </div>
      </div>
      <div css={styles.amoutArea.container}>
        <Heading size='h5' css={styles.amoutArea.text}>
          <span>Â¥</span> {amount.toLocaleString()}
        </Heading>
      </div>
    </Paper>
  )
}
