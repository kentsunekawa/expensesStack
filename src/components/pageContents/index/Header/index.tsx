// import from libraries
import 'styled-components/macro'
import { useCallback, useMemo } from 'react'
import { IconButton } from '@mui/material'
import {
  ArrowLeftRounded as ArrowLeftRoundedIcon,
  ArrowRightRounded as ArrowRightRoundedIcon,
} from '@mui/icons-material'
import { getYear, getMonth } from 'date-fns'

// import from this project
import { dateToString, getfirstDateOfAfterNthMonth } from 'src/utils'
import { useStyle } from 'src/hooks'
import { Heading } from 'src/components/parts/Texts'
import { createStyles } from './styles'

export type Props = {
  date: Date
  onChangeDate: (date: Date) => void
}

const today = new Date()

export const Header: React.FC<Props> = ({ date, onChangeDate }) => {
  const { styles } = useStyle(createStyles)

  const canGoNextMonth = useMemo(() => {
    const now = {
      year: getYear(today),
      month: getMonth(today),
    }

    const firstDateOfNextMonth = getfirstDateOfAfterNthMonth(date, 1)
    const next = {
      year: getYear(firstDateOfNextMonth),
      month: getMonth(firstDateOfNextMonth),
    }

    return next.year <= now.year && next.month <= now.month
  }, [date])

  const handleClickMoveButton = useCallback(
    (diff: -1 | 1) => {
      onChangeDate(getfirstDateOfAfterNthMonth(date, diff))
    },
    [date, onChangeDate],
  )

  return (
    <div css={styles.container}>
      <div css={styles.moveButtonArea}>
        <IconButton onClick={() => handleClickMoveButton(-1)}>
          <ArrowLeftRoundedIcon />
        </IconButton>
      </div>
      <div>
        <Heading size='h6'>{dateToString(date, 'yyyy, MMMM')}</Heading>
      </div>
      <div css={styles.moveButtonArea}>
        {canGoNextMonth && (
          <IconButton onClick={() => handleClickMoveButton(1)}>
            <ArrowRightRoundedIcon />
          </IconButton>
        )}
      </div>
    </div>
  )
}
