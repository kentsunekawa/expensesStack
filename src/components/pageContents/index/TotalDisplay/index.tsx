// import from libraries
import 'styled-components/macro'
import { useCallback, useEffect } from 'react'
import {
  FormControl,
  Select,
  InputLabel,
  MenuItem,
  SelectChangeEvent,
  Paper,
  Divider,
} from '@mui/material'

// import from this project
import { useStyle, useCategories } from 'src/hooks'
import { Suspense } from 'src/components/parts/Suspense'
import { Subtitle } from 'src/components/parts/Texts'
import { createStyles } from './styles'

export type Props = {
  category: {
    id: string
    name: string
  } | null
  onChangeCategory: (category: { id: string; name: string } | null) => void
  totalNum: string
}

export const TotalDisplay: React.FC<Props> = ({
  category,
  totalNum,
  onChangeCategory,
}) => {
  const { styles } = useStyle(createStyles)

  const { categories, fetchStatus, doGetCategories } = useCategories()

  const handleChangeCategory = useCallback(
    (e: SelectChangeEvent) => {
      if (categories) {
        onChangeCategory(
          categories.find(({ id }) => id === e.target.value) ?? null,
        )
      }
    },
    [categories, onChangeCategory],
  )

  useEffect(() => {
    doGetCategories()
  }, [doGetCategories])

  return (
    <Paper css={styles.container}>
      <div css={styles.categoryArea}>
        <Suspense
          {...fetchStatus}
          loadingProps={{
            size: 24,
          }}
        >
          <FormControl fullWidth size='small'>
            <InputLabel id='category'>Category</InputLabel>
            <Select
              labelId='category'
              value={category?.id ?? ''}
              label='Category'
              onChange={handleChangeCategory}
              size='small'
            >
              {categories?.map(({ id, name }) => (
                <MenuItem value={id} key={id}>
                  {name}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </Suspense>
      </div>
      <div css={styles.togalArea.container}>
        <Subtitle css={styles.togalArea.text} textAlign='right' size='large'>
          Â¥ {Number(totalNum).toLocaleString()}
        </Subtitle>
        <Divider css={styles.togalArea.divider} />
      </div>
    </Paper>
  )
}
