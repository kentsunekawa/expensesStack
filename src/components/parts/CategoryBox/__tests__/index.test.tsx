import { render, screen } from '@testing-library/react'

import { CategoryBox } from '../index'

describe('CategoryBox', () => {
  it('テキストが正しく表示される', () => {
    const name = 'category name'
    render(<CategoryBox name={name} />)
    screen.getByText(name)
  })
  it('背景色が正しく設定される', () => {
    const color = '#ddd'
    const { container } = render(<CategoryBox name='hogehoge' color={color} />)
    expect(container.querySelector('span')).toHaveStyle(
      `background: '${color}'`,
    )
  })

  it('Snapshot', () => {
    const { container } = render(<CategoryBox name='hogehoge' color={'#f00'} />)
    expect(container).toMatchSnapshot()
  })
})
