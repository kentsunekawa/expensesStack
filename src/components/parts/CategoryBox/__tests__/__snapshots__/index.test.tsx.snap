// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryBox Snapshot 1`] = `
<div>
  <div
    css="
    display: flex;
    gap: 0 4px;
    align-items: center;
  "
  >
    <span
      css="
    display: block;
    width: 16px;
    height: 16px;
    border-radius: 50%;
    margin-right: 4px;
    background: ,#f00,;
  "
    />
    <p
      class="MuiTypography-root-hjKMqV lcfbwq MuiTypography-root MuiTypography-body1"
      css="
      font-size: 14px;
      line-height: 1.6em;
      letter-spacing: 0;
    ,,"
    >
      hogehoge
    </p>
  </div>
</div>
`;
